### PR TITLE
Handle patterns in populate switch expressions

### LINKED ISSUE #50982: IDE0072: False positive when using switch expression and pattern matching with "or"
**Version Used**: 
.NET SDK 5.0.102

**Steps to Reproduce**:
`dotnet build` the following:
### Program.cs
```csharp
bool EnumIsValid(FriendOrFoe f)
{
    return f switch
    {
        FriendOrFoe.Friend or FriendOrFoe.Foe => true,
        _ => false,
    };
}

enum FriendOrFoe
{
    Friend,
    Foe
}
```

### ConsoleApp0.csproj
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net5.0</TargetFramework>
    <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
  </PropertyGroup>

</Project>
```

### .editorconfig
```editorconfig
[*.cs]
dotnet_diagnostic.IDE0072.severity = warning
```
**Expected Behavior**:
IDE0072 is not raised: all cases are handled

**Actual Behavior**:
IDE0072 is raised.  The suggested fix is to add a separate case for `FriendOrFoe.Friend` and a separate case for `FriendOrFor.Foe`.

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #50982

Not sure of the fix quality tbh.

### CODE REVIEW

File: src/Analyzers/Core/Analyzers/PopulateSwitch/PopulateSwitchExpressionHelpers.cs [hunk 1]
```diff
@@ -56,11 +56,52 @@ public static ICollection<ISymbol> GetMissingEnumMembers(ISwitchExpressionOperat
...
+                    if (constantValue.HasValue)
+                    {
+                        enumMembers.Remove(IntegerUtilities.ToInt64(constantValue.Value));
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+        }
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/PopulateSwitch/PopulateSwitchExpressionHelpers.cs:103]:
```suggestion
        private static void HandleBinaryPattern(IBinaryPatternOperation? binaryPattern, Dictionary<long, ISymbol> enumMembers)
        {
            if (binaryPattern?.OperatorKind == BinaryOperatorKind.Or)
            {
                RemoveIfConstantPatternHasValue(binaryPattern.LeftPattern);
                RemoveIfConstantPatternHasValue(binaryPattern.RightPattern);
                
                HandleBinaryPattern(binaryPattern.LeftPattern as IBinaryPatternOperation);
                HandleBinaryPattern(binaryPattern.RightPattern as IBinaryPatternOperation);
            }

            void RemoveIfConstantPatternHasValue(IOperation operation)
            {
                if (operation is IConstantPatternOperation { Value { ConstantValue { HasValue: true, Value: var value } } })
                        enumMembers.Remove(IntegerUtilities.ToInt64(value));
            }
        }
```

Comment (Youssef1313) [reply src/Analyzers/Core/Analyzers/PopulateSwitch/PopulateSwitchExpressionHelpers.cs:103]: @CyrusNajmabadi That's ALOT better! Loved how the code transformed into a much readable and neater code.

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Analyzers/Core/Analyzers/PopulateSwitch/PopulateSwitchExpressionHelpers.cs [hunk 2]
```diff
@@ -56,11 +56,33 @@ public static ICollection<ISymbol> GetMissingEnumMembers(ISwitchExpressionOperat
 
                     enumMembers.Remove(IntegerUtilities.ToInt64(constantValue.Value));
                 }
+                else if (arm.Pattern is IBinaryPatternOperation binaryPattern)
+                {
+                    HandleBinaryPattern(binaryPattern, enumMembers);
+                }
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/PopulateSwitch/PopulateSwitchExpressionHelpers.cs:62]: 1. do we need a 'false' here to handle all other types of pattern?
2. should HandleBinaryPattern return a bool to say if it couldn't handle this

It feels like we want this so we can appropriately bail out of the feature on things taht we don't understand.

File: src/Analyzers/Core/Analyzers/PopulateSwitch/PopulateSwitchExpressionHelpers.cs [hunk 3]
```diff
@@ -56,11 +56,33 @@ public static ICollection<ISymbol> GetMissingEnumMembers(ISwitchExpressionOperat
 
                     enumMembers.Remove(IntegerUtilities.ToInt64(constantValue.Value));
                 }
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/PopulateSwitch/PopulateSwitchExpressionHelpers.cs:58]: unify the code above this with the code in RemoveIfConstantPatternHasValue

File: src/Analyzers/Core/Analyzers/PopulateSwitch/PopulateSwitchExpressionHelpers.cs (hunk 2 above)
Comment (Youssef1313) [reply src/Analyzers/Core/Analyzers/PopulateSwitch/PopulateSwitchExpressionHelpers.cs:62]: @CyrusNajmabadi What are other patterns that can be relevant for enums?

I don't think we should bail out for things we don't understand. For example, if an arm has an "and" pattern, we still need to analyze other arms. See `TestMixingOrWithAndPatterns2`, for which I think the current behavior is correct.

Comment (CyrusNajmabadi) [reply src/Analyzers/Core/Analyzers/PopulateSwitch/PopulateSwitchExpressionHelpers.cs:62]: good point.  thanks.  but then why does the higher 'if' bail out on a value it doesn't understand.  feels like we shouidl at least be conssitent on that point wrt constant patterns.

Comment (Youssef1313) [reply src/Analyzers/Core/Analyzers/PopulateSwitch/PopulateSwitchExpressionHelpers.cs:51]: @CyrusNajmabadi Agreed. I updated it to be consistent.

File: src/Analyzers/CSharp/Tests/PopulateSwitch/PopulateSwitchExpressionTests.cs [hunk 4]
```diff
@@ -962,7 +962,7 @@ void Method()
             (MyEnum)0 => 1,
             (MyEnum)1 => 2,
             ""Mismatching constant"" => 3,
-            _ => throw new System.NotImplementedException(),
+            MyEnum.FizzBuzz => throw new System.NotImplementedException(),
```
Comment (Youssef1313) [inline src/Analyzers/CSharp/Tests/PopulateSwitch/PopulateSwitchExpressionTests.cs:965]: When "Mismatching constant" was encountered, the analyzer considered the switch as "complete". Hence, it failed to 'add missing cases'.

The "mismatching constant" is a compile-error anyway. So it probably doesn't matter a lot.

File: src/Analyzers/CSharp/Tests/PopulateSwitch/PopulateSwitchExpressionTests.cs [hunk 5]
```diff
@@ -1040,7 +1040,7 @@ void Main()
             Bar.Option1 => 1,
             Bar.Option2 => 2,
             null => null,
-            _ => throw new System.NotImplementedException(),
+            Bar.Option3 => throw new System.NotImplementedException(),
```
Comment (Youssef1313) [inline src/Analyzers/CSharp/Tests/PopulateSwitch/PopulateSwitchExpressionTests.cs:1043]: This test is misleading. It's named `TestAddMissingCasesForNullableEnum`, but the enum in it isn't nullable. Do you want me to make the enum nullable in it?

Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/Tests/PopulateSwitch/PopulateSwitchExpressionTests.cs:1043]: dont' change existing tests.  you can rename them and/or add new tests.  thanks!

Comment (Youssef1313) [reply src/Analyzers/CSharp/Tests/PopulateSwitch/PopulateSwitchExpressionTests.cs:1043]: @CyrusNajmabadi Well, I'll leave the test case and test name as-is for now. Only the expected result is updated.

Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/Tests/PopulateSwitch/PopulateSwitchExpressionTests.cs:1043]: wfm.  feel free to update the name as well if you want.  but genrally don't make the test fit the name, make the name fit the test. and we feel there is a test hole, where no test checks the original name's case, then we can add that :)

**Comment by CyrusNajmabadi** [comment]
Want me to merge?  or are you still working on things here?

**Comment by Youssef1313** [comment]
@CyrusNajmabadi You can merge.
Tests in this file may be updated in a follow-up. They need work to move to the new testing library, so that we can test different code actions individually (I assume there is no way to do that with the legacy testing infrastructure).

Let me know if you want me to file an issue for that?

**Comment by CyrusNajmabadi** [comment]
> Let me know if you want me to file an issue for that?

No issue needed.  We generally are not good about effectively managing such issues.  So the repo just fills up without end.

**Comment by Youssef1313** [comment]
@CyrusNajmabadi Anything left to merge?

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Pull request merged**